include/kernel.h

```diff
@@ -2056,6 +2057,76 @@ static inline void *z_impl_k_queue_peek_tail(struct k_queue *queue)
 
 /** @} */
 
+#ifdef CONFIG_USERSPACE
+/**
+ * @brief futex structure
+ *
+ * k_futex are tracked as kernel objects and can live in user memory
+ * so any access bypasses the kernel object permission management
+ * mechanism.
+ */
+struct k_futex {
+	atomic_t val;
+};
+
+/**
+ * @defgroup futex_apis FUTEX APIs
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/**
+ * @brief Initialize a futex.
+ *
+ * This routine initializes a futex object, prior to its first use.
+ *
+ * @param futex Address of the k_futex.
+ *
+ * @return N/A
+ */
+__syscall void k_futex_init(struct k_futex *futex);
+
+/**
+ * @brief Pend the current thread on a futex
+ *
+ * Tests that the supplied futex contains the expected value, and if so,
+ * goes to sleep until some other thread calls k_futex_wake() on it.
```

Comment:
Why does it go to sleep? Is it a formal expression to indicate that the thread is moved to pending state?

Reply:
@ioannisg you should probably review http://man7.org/linux/man-pages/man2/futex.2.html we are using the same semantics as Linux.